common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1373,23 +1354,41 @@ public void onSSLStateChange(int type, int val) {
                 case SSL_CB_HANDSHAKE_START: {
                     // For clients, this will allow the NEED_UNWRAP status to be
                     // returned.
-                    engineState = STATE_HANDSHAKE_STARTED;
+                    state = STATE_HANDSHAKE_STARTED;
                     break;
                 }
                 case SSL_CB_HANDSHAKE_DONE: {
-                    if (engineState != STATE_HANDSHAKE_STARTED
-                            && engineState != STATE_READY_HANDSHAKE_CUT_THROUGH) {
+                    if (state != STATE_HANDSHAKE_STARTED
+                            && state != STATE_READY_HANDSHAKE_CUT_THROUGH) {
                         throw new IllegalStateException(
-                                "Completed handshake while in mode " + engineState);
+                                "Completed handshake while in mode " + state);
                     }
-                    engineState = STATE_HANDSHAKE_COMPLETED;
+                    state = STATE_HANDSHAKE_COMPLETED;
                     break;
                 }
             }
         }
     }
 
     @Override
+    public boolean onNewSessionEstablished(long sslSessionNativePtr) {
+        try {
+            // Cache the newly established session.
+            AbstractSessionContext ctx = sessionContext();
+            ctx.cacheSession(SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession));
```

Comment:
@davidben Good catch! Yes I think moving `newInstance` to the first line is the way to go. I think I'd like to keep the signature to this method the same as the BoringSSL API, however for consistency.